lib/rules/no-constant-condition.js

```diff
@@ -143,11 +204,18 @@ module.exports = {
         //--------------------------------------------------------------------------
 
         return {
-            ConditionalExpression: checkConstantCondition,
-            IfStatement: checkConstantCondition,
+            ConditionalExpression: reportIfConstant,
+            IfStatement: reportIfConstant,
             WhileStatement: checkLoop,
+            "WhileStatement:exit": checkConstantConditionLoopInSet,
             DoWhileStatement: checkLoop,
-            ForStatement: checkLoop
+            "DoWhileStatement:exit": checkConstantConditionLoopInSet,
+            ForStatement: checkLoop,
+            "ForStatement > .init * > YieldExpression, ForStatement > .init": node => checkYieldInInitForLoop(node),
```

Comment:
The following should work:

<details>
<summary>View diff</summary>

```diff
diff --git a/lib/rules/no-constant-condition.js b/lib/rules/no-constant-condition.js
index 9f38c790..e666c776 100644
--- a/lib/rules/no-constant-condition.js
+++ b/lib/rules/no-constant-condition.js
@@ -187,18 +187,6 @@ module.exports = {
             }
         }
 
-        /**
-        * Checks node for yield in init of forloop
-        * @param {ASTNode} node The AST node to check.
-        * @returns {void}
-        * @private
-        */
-        function checkYieldInInitForLoop(node) {
-            if (node.type === "YieldExpression" || (node.right && node.right.type === "YieldExpression")) {
-                context.report({ node: node.parent, message: "Unexpected constant condition." });
-            }
-        }
-
         //--------------------------------------------------------------------------
         // Public
         //--------------------------------------------------------------------------
@@ -210,8 +198,7 @@ module.exports = {
             "WhileStatement:exit": checkConstantConditionLoopInSet,
             DoWhileStatement: checkLoop,
             "DoWhileStatement:exit": checkConstantConditionLoopInSet,
-            ForStatement: checkLoop,
-            "ForStatement > .init * > YieldExpression, ForStatement > .init": node => checkYieldInInitForLoop(node),
+            "ForStatement > .test": node => checkLoop(node.parent),
             "ForStatement:exit": checkConstantConditionLoopInSet,
             FunctionDeclaration: enterFunction,
             "FunctionDeclaration:exit": exitFunction,

```
</details>

I think the confusion happened because I accidentally asked you to add some incorrect tests. I was under the impression that the rule was supposed to report things like `for(; ; );` (i.e. for loops that are missing a `test` property), but I was wrong -- sorry about that.

When testing this locally, I made the following corrections to the tests, which made them all pass:

<details>
<summary>View diff</summary>

```diff
diff --git a/tests/lib/rules/no-constant-condition.js b/tests/lib/rules/no-constant-condition.js
index 312f04f5..8897b340 100644
--- a/tests/lib/rules/no-constant-condition.js
+++ b/tests/lib/rules/no-constant-condition.js
@@ -66,7 +66,10 @@ ruleTester.run("no-constant-condition", rule, {
         { code: "function* foo(){while (true) { while(true) {yield;}}}" },
         { code: "function* foo() {for (; yield; ) {}}" },
         { code: "function* foo() {for (; ; yield) {}}" },
-        { code: "function* foo() {while (true) {function* foo() {yield;}yield;}}" }
+        { code: "function* foo() {while (true) {function* foo() {yield;}yield;}}" },
+
+        "function* foo() { for (let x = yield; x < 10; x++); }",
+        "function* foo() { for (let x = yield; ; x++) { yield; } }"
     ],
     invalid: [
         { code: "for(;true;);", errors: [{ message: "Unexpected constant condition.", type: "ForStatement" }] },
@@ -139,11 +142,11 @@ ruleTester.run("no-constant-condition", rule, {
             errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }]
         },
         {
-            code: "function* foo() {for (let foo = yield; ;) {}}",
-            errors: [{ message: "Unexpected constant condition.", type: "VariableDeclarator" }]
+            code: "function* foo() {for (let foo = yield; true;) {}}",
+            errors: [{ message: "Unexpected constant condition.", type: "ForStatement" }]
         },
         {
-            code: "function* foo() {for (foo = yield; ;) {}}",
+            code: "function* foo() {for (foo = yield; true;) {}}",
             errors: [{ message: "Unexpected constant condition.", type: "ForStatement" }]
         },
         {
@@ -151,8 +154,8 @@ ruleTester.run("no-constant-condition", rule, {
             errors: [{ message: "Unexpected constant condition.", type: "WhileStatement" }]
         },
         {
-            code: "function* foo() { for (let foo = 1 + 2 + 3 + (yield); bar; baz) {} }",
-            errors: [{ message: "Unexpected constant condition.", type: "BinaryExpression" }]
+            code: "function* foo() { for (let foo = 1 + 2 + 3 + (yield); true; baz) {} }",
+            errors: [{ message: "Unexpected constant condition.", type: "ForStatement" }]
         }
     ]
 });

```

</details>